### ISSUE TITLE
GDFError: b'GDF_UNSUPPORTED_DTYPE'

### ISSUE THREAD

**Issue by tonystarkiss** [rapidsai/cudf#403]
## Reporting a bug

- [x] I am using the latest version of cuDF from conda, built from `master`, or
built from the latest tagged release. 
- [ ] I have included the version or commit hash I am using for reference.
- [ ] I have included the following environment details:
      Linux Distro, Linux Kernel, GPU Model
- [ ] I have included the following version information for:
      Arrow, cmake, CUDA, gcc/g++, Numpy, Pandas, Python
- [ ] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

this is the python 3.5 code below ,with pandas 0.20.3, the newest cuDF , there are 2 examples in the code below , the first example with the N value of 126 , it will run success , the second example with the N value of 127 , it failed .

```
import pandas as pd
import cudf as cuDF
import datetime
from time import time
import random, string

def randomword(length):
    letters = string.ascii_lowercase
    return ''.join(random.choice(letters) for i in range(length))

#Success example:
len_vector=5
N = 10000

pet_owner = list(range(N))
pet_type = [ 'dog', 'fish', 'cat', 'plane', 'bird']*int(N/len_vector)
df = pd.DataFrame({'pet_owner': pet_owner, 'pet_type': pet_type})
df.pet_type = df.pet_type.astype('category')
my_gdf = cuDF.DataFrame.from_pandas(df)
my_gdf['pet_codes'] = my_gdf.pet_type.cat.codes
codes = my_gdf.pet_codes.unique()
enc_gdf = my_gdf.one_hot_encoding('pet_codes', 'pet_dummy', codes)
enc_df = enc_gdf.to_pandas()
print(enc_df)


len_vector=5
N = 126


print("Example 1")
pet_owner = list(range(N))
pet_type2 = [randomword(len_vector) for _ in range(N)]

print(pet_type2)
df_2 = pd.DataFrame({'pet_owner': pet_owner, 'pet_type': pet_type2})
df_2.pet_type = df_2.pet_type.astype('category')

print(df_2)

start_time=datetime.datetime.fromtimestamp(time()).strftime('%Y-%m-%d %H:%M:%S')

my_gdf2 = cuDF.DataFrame.from_pandas(df_2)
my_gdf2['pet_codes'] = my_gdf2.pet_type.cat.codes
codes = my_gdf2.pet_codes.unique()
enc_gdf2 = my_gdf2.one_hot_encoding('pet_codes', 'pet_dummy', codes)

enc_df2 = enc_gdf2.to_pandas()
end_time = datetime.datetime.fromtimestamp(time()).strftime('%Y-%m-%d %H:%M:%S')
print(enc_df2)


print("start_time = {}".format(start_time))
print("end_time = {}".format(end_time))


len_vector=5
N = 127


print("Example 2")
pet_owner = list(range(N))
pet_type3 = [randomword(len_vector) for _ in range(N)]

print(pet_type3)
df_3 = pd.DataFrame({'pet_owner': pet_owner, 'pet_type': pet_type3})
df_3.pet_type = df_3.pet_type.astype('category')

print(df_3)

start_time=datetime.datetime.fromtimestamp(time()).strftime('%Y-%m-%d %H:%M:%S')

my_gdf3 = cuDF.DataFrame.from_pandas(df_3)
my_gdf3['pet_codes'] = my_gdf3.pet_type.cat.codes
codes = my_gdf3.pet_codes.unique()
enc_gdf3 = my_gdf3.one_hot_encoding('pet_codes', 'pet_dummy', codes)

enc_df3 = enc_gdf3.to_pandas()
end_time = datetime.datetime.fromtimestamp(time()).strftime('%Y-%m-%d %H:%M:%S')
print(enc_df3)


print("start_time = {}".format(start_time))
print("end_time = {}".format(end_time))
```

=============================================
when the value of N is larger than 126 , such as 127 at the second example , it will encounter the error below .

```
GDFError                                  Traceback (most recent call last)
<ipython-input-8-2976d3914212> in <module>
     19 my_gdf3 = cuDF.DataFrame.from_pandas(df_3)
     20 my_gdf3['pet_codes'] = my_gdf3.pet_type.cat.codes
---> 21 codes = my_gdf3.pet_codes.unique()
     22 enc_gdf3 = my_gdf3.one_hot_encoding('pet_codes', 'pet_dummy', codes)
     23 

/conda/lib/python3.5/site-packages/cudf/dataframe/series.py in unique(self, method, sort)
    806         if self.null_count == len(self):
    807             return np.empty(0, dtype=self.dtype)
--> 808         res = self._column.unique(method=method)
    809         return Series(res)
    810 

/conda/lib/python3.5/site-packages/cudf/dataframe/numerical.py in unique(self, method)
    180             msg = 'non sort based unique() not implemented yet'
    181             raise NotImplementedError(msg)
--> 182         segs, sortedvals = self._unique_segments()
    183         # gather result
    184         out = cudautils.gather(data=sortedvals, index=segs)

/conda/lib/python3.5/site-packages/cudf/dataframe/numerical.py in _unique_segments(self)
    168         densecol = self.replace(data=self.to_dense_buffer(), mask=None)
    169         # sort the column
--> 170         sortcol, _ = densecol.sort_by_values(ascending=True)
    171         # find segments
    172         sortedvals = sortcol.to_gpu_array()

/conda/lib/python3.5/site-packages/cudf/dataframe/numerical.py in sort_by_values(self, ascending)
    137         inds = Buffer(cudautils.arange(len(self)))
    138         col_inds = self.replace(data=inds, dtype=inds.dtype)
--> 139         cpp_sort.apply_sort(col_keys, col_inds, ascending=ascending)
    140         return col_keys, col_inds
    141 

cudf/bindings/sort.pyx in cudf.bindings.sort.apply_sort()

cudf/bindings/sort.pyx in cudf.bindings.sort.apply_sort()

cudf/bindings/cudf_cpp.pyx in cudf.bindings.cudf_cpp.check_gdf_error()

GDFError: b'GDF_UNSUPPORTED_DTYPE'
```

**Comment by kkraus14** [comment]
Looks like we're using `int8` for the integer representation of categoricals instead of properly assigning a type based on the number of categories.

**Comment by shwina** [comment]
@kkraus14 I think the categorical types are handled correctly. The problem is that int16 types are not supported by sort:

https://github.com/rapidsai/cudf/blob/master/cpp/src/sort/sorting.cu#L208

I can confirm that adding a `gdf_radixsort_i16` fixes this issue. If that sounds like the right approach, I'm happy to submit a PR!

**Comment by kkraus14** [comment]
@shwina I believe this will be resolved by PR #411 and Python bindings surrounding that revised libcudf order_by function.

**Comment by kkraus14** [comment]
This should be resolved by #411.
